pkg/network/ebpf/c/runtime/http.c

```diff
@@ -22,13 +22,51 @@
 static __always_inline void read_into_buffer_skb(char *buffer, struct __sk_buff* skb, skb_info_t *info) {
     u64 offset = (u64)info->data_off;
 
+#define BLK_SIZE (16)
+    const u32 iter = HTTP_BUFFER_SIZE / BLK_SIZE;
+    const u32 len = HTTP_BUFFER_SIZE < (skb->len - (u32)offset) ? (u32)offset + HTTP_BUFFER_SIZE : skb->len;
+
+    unsigned i = 0;
+
 #pragma unroll
-    for (int i = 0; i < HTTP_BUFFER_SIZE; i++) {
-        if (offset < skb->len) {
-            bpf_skb_load_bytes(skb, offset, &buffer[i], 1);
-        }
-        offset++;
+    for (; i < iter; i++) {
+        if (offset + BLK_SIZE - 1 >= len) break;
+
+        bpf_skb_load_bytes(skb, offset, &buffer[i * BLK_SIZE], BLK_SIZE);
+        offset += BLK_SIZE;
     }
+
+    void *buf = &buffer[i * BLK_SIZE];
+    if (offset + 14 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 15);
+    else if (offset + 13 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 14);
+    else if (offset + 12 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 13);
+    else if (offset + 11 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 12);
+    else if (offset + 10 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 11);
+    else if (offset + 9 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 10);
+    else if (offset + 8 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 9);
+    else if (offset + 7 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 8);
+    else if (offset + 6 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 7);
+    else if (offset + 5 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 6);
+    else if (offset + 4 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 5);
+    else if (offset + 3 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 4);
+    else if (offset + 2 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 3);
+    else if (offset + 1 < len)
+        bpf_skb_load_bytes(skb, offset, buf, 2);
+    else if (offset < len)
+        bpf_skb_load_bytes(skb, offset, buf, 1);
```

Comment:
Can you explain why this big else-if block is necessary? Is it to satisfy a certain kernel version's verifier?

Reply:
Yep, this is basically the same explanation as the prebuilt version, I'll add a comment about that here as well!